src/aeso_fcode_to_fate.erl

```diff
@@ -179,42 +182,45 @@ lit_to_fate(L) ->
         {typerep, T}         -> aeb_fate_data:make_typerep(type_to_scode(T))
      end.
 
-term_to_fate({lit, L}) ->
+term_to_fate(E) -> term_to_fate(#{}, E).
+
+term_to_fate(_Env, {lit, L}) ->
     lit_to_fate(L);
 %% negative literals are parsed as 0 - N
-term_to_fate({op, '-', [{lit, {int, 0}}, {lit, {int, N}}]}) ->
+term_to_fate(_Env, {op, '-', [{lit, {int, 0}}, {lit, {int, N}}]}) ->
     aeb_fate_data:make_integer(-N);
-term_to_fate(nil) ->
+term_to_fate(_Env, nil) ->
     aeb_fate_data:make_list([]);
-term_to_fate({op, '::', [Hd, Tl]}) ->
+term_to_fate(Env, {op, '::', [Hd, Tl]}) ->
     %% The Tl will translate into a list, because FATE lists are just lists
-    [term_to_fate(Hd) | term_to_fate(Tl)];
-term_to_fate({tuple, As}) ->
-    aeb_fate_data:make_tuple(list_to_tuple([ term_to_fate(A) || A<-As]));
-term_to_fate({con, Ar, I, As}) ->
-    FateAs = [ term_to_fate(A) || A <- As ],
+    [term_to_fate(Env, Hd) | term_to_fate(Env, Tl)];
+term_to_fate(Env, {tuple, As}) ->
```

Comment:
Possibly I do not really get all type layers, but {tuple, [A]} in this case would not translate to just A as in scode, but rather keep the FATE tuple notation.

Most likely on purpose.

Reply:
No, you're right. That probably needs a case for singletons.

Reply:
I'll try to make a test case.

Reply:
No, it's right as it is. The singleton optimisation happens in `ast_to_fcode` and `term_to_fate` works on `fcode` so it's already happened at that point.